src/tiered/tiered_handle.c

```diff
@@ -8,6 +8,67 @@
 
 #include "wt_internal.h"
 
+/*
+ * __tiered_name_check --
+ *     Check if the given tiered table name exists in the shared storage. This is called from create
+ *     to check if this name had previously existed, was dropped and we're now trying to create the
+ *     same name. We cannot do that if shared objects exist for the old table.
+ */
+static int
+__tiered_name_check(WT_SESSION_IMPL *session, WT_TIERED *tiered)
+{
+    WT_DECL_RET;
+    WT_FILE_SYSTEM *bucket_fs;
+    size_t len, obj_len;
+    u_int obj_count, i;
+    char **obj_files;
+    const char *name, *obj_name, *obj_uri;
+
+    bucket_fs = tiered->bstorage->file_system;
+    name = tiered->iface.name;
+    obj_name = obj_uri = NULL;
+    WT_ASSERT(session, WT_PREFIX_MATCH(name, "tiered:"));
+    WT_PREFIX_SKIP(name, "tiered:");
+    /* See if this name exists in the shared storage. */
+    __wt_verbose(session, WT_VERB_TIERED, "NAME_CHECK: check for %s", name);
+    WT_RET(bucket_fs->fs_directory_list(bucket_fs, (WT_SESSION *)session,
+      tiered->bstorage->bucket_prefix, name, &obj_files, &obj_count));
+    __wt_verbose(session, WT_VERB_TIERED, "NAME_CHECK: Got %d files", (int)obj_count);
+    if (obj_count == 0)
+        goto done;
+    /*
+     * We need to distinguish this name from a superset name so check the length also matches.
+     * Generate an object name so that we know the maximum length that our name should be and we can
+     * check for the right form.
+     */
+    WT_ERR(__wt_tiered_name(session, &tiered->iface, 1, WT_TIERED_NAME_OBJECT, &obj_uri));
+    obj_name = obj_uri;
+    WT_PREFIX_SKIP_REQUIRED(session, obj_name, "object:");
+    /* This is the length of the name-<object number>.wtobj string. */
+    obj_len = strlen(obj_name);
```

Comment:
Consider calling `bucket_fs->fs_directory_list` at this point, using `obj_name` instead of `name`.  That should give a tiny list (if any) and save a lot of processing - especially when we get to thousands of objects.

Reply:
I am not following this. I do call `bucket_fs->fs_directory_list` on line 34. And that is the tiny list I'm comparing against. Or I think I am. Can you please let me know what I'm missing here?

Reply:
I think misunderstood what the function does - so disregard that comment.  Just so I understand, it looks like if we are talking about `table:foo`, then (forgetting about prefixes), we're looking for names that match `foo-0.wtobj`, `foo-1.wtobj` etc.?  So you generate `foo-1.wtobj` and list matches for "foo" and see if any of the lengths of the result match length of `foo-1.wtobj`?  But what if (after GC is implemented in the future), lower numbers have been removed and there exists only `foo-25.wtobj` through`foo-35.wtobj`?  So I think the matching has to be more sophisticated than just a length match.

Reply:
The length is zero-filled and fixed for a name `foo`. It will be `foo-00001.wtobj, foo-00025.wtobj, foo-00321.wtobj` etc. (There are a lot more zeroes in the real definition). I was using the length plus name `foo` so that it does not match `fo` or `food` or even a table named `foo-00001`.

Reply:
Right, I forgot that.  Might be worth adding a few words reminding that the object name suffixes are fixed length.

Reply:
Fixed.